aspnetcore/fundamentals/startup.md

```diff
@@ -65,7 +65,57 @@ Each `Use` extension method adds a [middleware](xref:fundamentals/middleware) co
 
 For more information about how to use `IApplicationBuilder`, see [Middleware](xref:fundamentals/middleware).
 
-Additional services, like `IHostingEnvironment` and `ILoggerFactory` may also be specified in the method signature, in which case these services will be [injected](dependency-injection.md) if they are available. 
+Additional services, like `IHostingEnvironment` and `ILoggerFactory` may also be specified in the method signature, in which case these services will be [injected](dependency-injection.md) if they are available.
+
+## Hosting startup light up feature
+
+An [IHostingStartup](/dotnet/api/microsoft.aspnetcore.hosting.ihostingstartup) implementation allows adding features to an app at startup outside of the app's `Startup` class. For example, tooling can use an `IHostingStartup` implementation to provide additional configuration providers or services to an app on startup.
+
+The [HostingStartupSample app](https://github.com/aspnet/Docs/tree/master/aspnetcore/fundamentals/startup/sample/) ([how to download](xref:tutorials/index#how-to-download-a-sample)) uses `IHostingStartup` to configure [options](xref:fundamentals/configuration/options) in the app's service container.
+
+The sample app's `StartupInjection` class implements `IHostingStartup`. The [Configure](/dotnet/api/microsoft.aspnetcore.hosting.ihostingstartup.configure) method uses an [IWebHostBuilder](/dotnet/api/microsoft.aspnetcore.hosting.iwebhostbuilder) to add an `InjectedOptions` class to the app's service container with assigned values. The [HostingStartupAttribute](/dotnet/api/microsoft.aspnetcore.hosting.hostingstartupattribute) identifies the `StartupInjection` class as an implementation of `IHostingStartup` for loading and execution when building the [IWebHost](/dotnet/api/microsoft.aspnetcore.hosting.iwebhost):
+
+[!code-csharp[Main](startup/sample/StartupInjection.cs?name=snippet1&highlight=1)]
+
+The sample app's Index page injects the options directly into the page for rendering:
+
+[!code-cshtml[Main](startup/sample/Pages/Index.cshtml?highlight=3-4,19-20,23-24)]
+
+When the app is run, the Index page displays the values:
+
+![Browser window showing the rendered Index page. The values of Option1 and Option2 are displayed as 'Option 1 Value' and 'Option 2 Value'.](startup/_static/index.png)
+
+## Startup filters
+
+Use [IStartupFilter](/dotnet/api/microsoft.aspnetcore.hosting.istartupfilter) to configure and control the order of middleware at the beginning or end of an app's [Configure](#the-configure-method) middleware pipeline.
```

Comment:
You don't get much control over ordering, you can only add at the start or end.

Reply:
I thought you can order the `IStartupFilter`s by the order of the service registrations, which would in-turn order the middlewares of each `IStartupFilter`.

Reply:
For example, say I had three of these. If each configured one middleware, would the middlewares defined by these run in-order 3-1-2?

```csharp
services.AddTransient<IStartupFilter, RequestSetOptionsStartupFilter3>();
services.AddTransient<IStartupFilter, RequestSetOptionsStartupFilter1>();
services.AddTransient<IStartupFilter, RequestSetOptionsStartupFilter2>();
services.AddMvc();
```

Reply:
That's app control of the order, the filter itself has no control.

Reply:
Ok, I just meant that when multiple `IStartupFilter` instances are there that they can be ordered by their service registrations. The concept that I'm *failing* :cry: to convey is that when libs with their own `IStartupFilter`s are registered the dev can order all of the middlewares to their liking (app-configured startup filters and lib-configured startup filters) ... that's the order I was referring to. I'll see if I can fix it on the next pass.